src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,652 @@
+        liquidityProviderController = _liquidityProviderController;
+
+        emit LiquidityProviderControllerUpdated(_liquidityProviderController);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Performance Fee Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Owner sets the performance fee on increased assets
+    /// @param _fee The performance fee measured in basis points (1/100th of a percent)
+    /// 10,000 = 100% performance fee
+    /// 500 = 5% performance fee
+    function setFee(uint256 _fee) external onlyOwner {
+        _setFee(_fee);
+    }
+
+    /// @notice Owner sets the account/contract that receives the performance fee
+    function setFeeCollector(address _feeCollector) external onlyOwner {
+        _setFeeCollector(_feeCollector);
+    }
+
+    function _setFee(uint256 _fee) internal {
+        require(_fee <= FEE_SCALE, "ARM: fee too high");
+
+        // Collect any performance fees up to this point using the old fee
+        collectFees();
+
+        fee = SafeCast.toUint16(_fee);
+
+        emit FeeUpdated(_fee);
+    }
+
+    function _setFeeCollector(address _feeCollector) internal {
+        require(_feeCollector != address(0), "ARM: invalid fee collector");
+
+        feeCollector = _feeCollector;
+
+        emit FeeCollectorUpdated(_feeCollector);
+    }
+
+    /// @notice Transfer accrued performance fees to the fee collector
+    /// This requires enough liquidity assets in the ARM to cover the accrued fees.
+    function collectFees() public returns (uint256 fees) {
+        uint256 newAvailableAssets;
+        // Accrue any performance fees up to this point
+        (fees, newAvailableAssets) = _feesAccrued();
+
+        if (fee == 0) return fees;
+
+        require(fees <= IERC20(liquidityAsset).balanceOf(address(this)), "ARM: insufficient liquidity");
+
+        IERC20(liquidityAsset).transfer(feeCollector, fees);
+
+        // Save the new available assets back to storage less the collected fees.
+        lastAvailableAssets = SafeCast.toInt128(SafeCast.toInt256(newAvailableAssets) - SafeCast.toInt256(fees));
+
+        emit FeeCollected(feeCollector, fees);
+    }
+
+    /// @notice Calculates the performance fees accrued since the last time fees were collected
+    function feesAccrued() public view returns (uint256 fees) {
+        (fees,) = _feesAccrued();
+    }
+
+    function _feesAccrued() public view returns (uint256 fees, uint256 newAvailableAssets) {
+        newAvailableAssets = _availableAssets();
+
+        // Calculate the increase in assets since the last time fees were calculated
+        int256 assetIncrease = SafeCast.toInt256(newAvailableAssets) - lastAvailableAssets;
+
+        // Do not accrued a performance fee if the available assets has decreased
+        if (assetIncrease <= 0) return (0, newAvailableAssets);
```

Comment:
🟠 This line creates two different yield rates depending on if we are above or below the previous high watermark. This slightly privileges mints made below high watermarks. Maybe we want this? Maybe not?

Consider totalAssets(), the most important use of this function's output, since it drives mints / redeems and all user funds.

```
(uint256 fees, uint256 newAvailableAssets) = _feesAccrued();
return newAvailableAssets - fees
```

As the contract's total assets go up, the users assets go up to, but not as fast - this difference is the fee.

When the assets go down, the user assets go down, also not as fast - this too is a part of the fee.

However, when we have just collected, and there is not a stash of fees sitting around, loss of assets makes the user assets go down 1:1 with the rate of assets going down. This also means that if anyone joins the pool here, they pay no fees until we re-reach the watermark price. That may not be completely bad, but just feels wrong.

I guess this keeps the protocol always solvent. Probably something to that.

![Drawing 14](https://github.com/user-attachments/assets/67f70199-4e9e-4e5c-adab-100fdadb2d76)

Reply:
I've reduced `MAX_PRICE_DEVIATION` to 2 basis points for now. That means the sell price and only go as low as 2 basis points below 1. ie 0.9998

Reply:
I've also made it so the buy price can not be 1 or more. So only the sell price can cross 1 and it can only cross by 2 basis points

Reply:
I think the current behavior is correct.